src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs

```diff
@@ -80,9 +80,13 @@ public static SourceText From(ITextBufferCloneService textBufferCloneServiceOpt,
 
                 if (!s_textSnapshotMap.TryGetValue(editorSnapshot, out var snapshot))
                 {
+                    // Explicitly obtain the TextBufferContainer before calling GetValue to avoid reentrancy in
+                    // ConditionalWeakTable. https://github.com/dotnet/roslyn/issues/28256
+                    var container = TextBufferContainer.From(editorSnapshot.TextBuffer);
+
                     // Avoid capturing `textBufferCloneServiceOpt` on the fast path
                     var tempTextBufferCloneServiceOpt = textBufferCloneServiceOpt;
-                    snapshot = s_textSnapshotMap.GetValue(editorSnapshot, s => new SnapshotSourceText(tempTextBufferCloneServiceOpt, s));
+                    snapshot = s_textSnapshotMap.GetValue(editorSnapshot, s => new SnapshotSourceText(tempTextBufferCloneServiceOpt, s, container));
```

Comment:
do we still need the version of the constructor that doesn't take a container?

Reply:
looking at current roslyn, it appears as if the answer is 'no'.  so i would remove that constructor if possible.